src/lem/circuit.rs

```diff
@@ -86,69 +84,81 @@ pub struct SlotWitness<F: LurkField> {
 
 /// Manages global allocations for constants in a constraint system
 #[derive(Default)]
-pub struct GlobalAllocator<F: LurkField>(HashMap<FWrap<F>, AllocatedNum<F>>);
+pub struct GlobalAllocator<F: LurkField>(FrozenMap<FWrap<F>, Box<AllocatedNum<F>>>);
 
 impl<F: LurkField> GlobalAllocator<F> {
-    /// Checks if the allocation for a numeric variable has already been cached.
-    /// If so, don't do anything. Otherwise, allocate and cache it.
-    pub fn new_const<CS: ConstraintSystem<F>>(&mut self, cs: &mut CS, f: F) {
-        self.0
-            .entry(FWrap(f))
-            .or_insert_with(|| allocate_constant(&mut cs.namespace(|| f.hex_digits()), f));
-    }
-
-    #[inline]
-    pub fn new_const_from_tag<CS: ConstraintSystem<F>, T: Tag>(&mut self, cs: &mut CS, tag: &T) {
-        self.new_const(cs, tag.to_field());
-    }
-
-    #[inline]
-    pub fn new_consts_from_z_ptr<CS: ConstraintSystem<F>>(&mut self, cs: &mut CS, z_ptr: &ZPtr<F>) {
-        self.new_const_from_tag(cs, z_ptr.tag());
-        self.new_const(cs, *z_ptr.value());
+    /// Memoizes allocations for numerical constants in a constraint system
+    pub fn alloc_const<CS: ConstraintSystem<F>>(&self, cs: &mut CS, f: F) -> &AllocatedNum<F> {
+        let key = FWrap(f);
+        if let Some(allocated_const) = self.0.get(&key) {
+            if let Some(allocated_f) = allocated_const.get_value() {
+                assert_eq!(allocated_f, f);
+            }
+            allocated_const
+        } else {
+            let allocated_const = allocate_constant(&mut cs.namespace(|| f.hex_digits()), f);
+            self.0.insert(key, allocated_const.into())
+        }
     }
 
     #[inline]
-    pub fn get_const(&self, f: F) -> Result<&AllocatedNum<F>, SynthesisError> {
-        self.0
-            .get(&FWrap(f))
-            .ok_or_else(|| SynthesisError::AssignmentMissing)
+    pub fn alloc_const_cloned<CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        f: F,
+    ) -> AllocatedNum<F> {
+        self.alloc_const(cs, f).clone()
     }
 
     #[inline]
-    pub fn get_const_cloned(&self, f: F) -> Result<AllocatedNum<F>, SynthesisError> {
-        self.get_const(f).cloned()
+    pub fn alloc_tag<CS: ConstraintSystem<F>, T: Tag>(
+        &self,
+        cs: &mut CS,
+        tag: &T,
+    ) -> &AllocatedNum<F> {
+        self.alloc_const(cs, tag.to_field())
     }
 
     #[inline]
-    pub fn get_tag<T: Tag>(&self, tag: &T) -> Result<&AllocatedNum<F>, SynthesisError> {
-        self.get_const(tag.to_field())
+    pub fn alloc_tag_cloned<CS: ConstraintSystem<F>, T: Tag>(
```

Comment:
Nit: for the record, I have so many issues with those aliases. The only thing that's worse than a code base littered with `.clone()` is one that's littered with invisible `.clone()`s.

(please ignore, it's not the point of this PR)